From 2e94ffec6aad40f747c16815d124cb8f96ad5873 Mon Sep 17 00:00:00 2001
From: Jeffy Chen <jeffy.chen@rock-chips.com>
Date: Tue, 9 Jul 2019 19:39:47 +0800
Subject: [PATCH 34/37] HACK: glamor: Use copy/put bail funcs for 64x64 images
 for freerdp

Freerdp would split screen into 64x64 images.
For small images, the glamor's performance is worse.

But preparing accessing requires buffer sync, so let's only hack the
64x64 images for freerdp case.

Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 glamor/glamor_copy.c  | 6 ++++++
 glamor/glamor_image.c | 6 ++++++
 2 files changed, 12 insertions(+)

diff --git a/glamor/glamor_copy.c b/glamor/glamor_copy.c
index 1ab2be6c0..82ae20dd1 100644
--- a/glamor/glamor_copy.c
+++ b/glamor/glamor_copy.c
@@ -701,6 +701,12 @@ glamor_copy_gl(DrawablePtr src,
     glamor_pixmap_private *src_priv = glamor_get_pixmap_private(src_pixmap);
     glamor_pixmap_private *dst_priv = glamor_get_pixmap_private(dst_pixmap);
 
+#ifdef GLAMOR_HAS_GBM_MAP
+    /* HACK for freerdp */
+    if (nbox == 1 && (box->x2 - box->x1) == 64 && (box->y2 - box->y1) == 64)
+        return FALSE;
+#endif
+
     if (GLAMOR_PIXMAP_PRIV_HAS_FBO(dst_priv)) {
         if (GLAMOR_PIXMAP_PRIV_HAS_FBO(src_priv)) {
             if (glamor_copy_needs_temp(src, dst, box, nbox, dx, dy))
diff --git a/glamor/glamor_image.c b/glamor/glamor_image.c
index 453ef79ba..129ca3051 100644
--- a/glamor/glamor_image.c
+++ b/glamor/glamor_image.c
@@ -41,6 +41,12 @@ glamor_put_image_gl(DrawablePtr drawable, GCPtr gc, int depth, int x, int y,
     BoxRec      box;
     int         off_x, off_y;
 
+#ifdef GLAMOR_HAS_GBM_MAP
+    /* HACK for freerdp */
+    if (w == 64 && h == 64)
+        return FALSE;
+#endif
+
     pixmap_priv = glamor_get_pixmap_private(pixmap);
 
     if (!GLAMOR_PIXMAP_PRIV_HAS_FBO(pixmap_priv))
-- 
2.11.0

